Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,605 @@
+        if (right->_mantissa[index] > left->_mantissa[index]) {
+            return NSOrderedAscending;
+        }
+
+        if (right->_mantissa[index] < left->_mantissa[index]) {
+            return NSOrderedDescending;
+        }
+    }
+
+    return NSOrderedSame;
+}
+
+static NSCalculationError _handleDifferenceOfSignAddition(NSDecimal* result,
+                                                          NSDecimal* left,
+                                                          NSDecimal* right,
+                                                          NSRoundingMode roundingMode) {
+    // Note: All inputs have been normalized via NSDecimalNormalize
+    // Since these are normalized and to account for unsigned subtraction we have to do mantissa based comparison.
+
+    NSComparisonResult res = _compareMantissa(left, right);
+
+    if (res == NSOrderedSame) {
+        SET_TO_ZERO(result);
+        return NSCalculationNoError;
+    }
+
+    if (res == NSOrderedAscending) {
+        // R > L
+        NSCalculationError error = _subtraction(result, right, left);
+        if (error != NSCalculationNoError) {
+            INVALIDATE_DECIMAL(result);
+            return error;
+        }
+        result->_isNegative = right->_isNegative;
+        result->_exponent = right->_exponent;
+    } else {
+        NSCalculationError error = _subtraction(result, left, right);
+        if (error != NSCalculationNoError) {
+            INVALIDATE_DECIMAL(result);
+            return error;
+        }
+        result->_isNegative = left->_isNegative;
+        result->_exponent = left->_exponent;
+    }
+
+    NSDecimalCompact(result);
+    return NSCalculationNoError;
+}
+
+/**
+ @Status Caveat
+ @Notes roundingMode is not supported.
+*/
+NSCalculationError NSDecimalAdd(NSDecimal* result,
+                                const NSDecimal* leftOperand,
+                                const NSDecimal* rightOperand,
+                                NSRoundingMode roundingMode) {
+    THROW_NS_IF_FALSE(E_INVALIDARG, (leftOperand != NULL) && (rightOperand != NULL) && (result != NULL));
+    VALIDATE_ARITHMETIC_INPUT(result, leftOperand, rightOperand);
+
+    if (IS_ZERO(leftOperand)) {
+        NSDecimalCopy(result, rightOperand);
+        return NSCalculationNoError;
+    }
+
+    if (IS_ZERO(rightOperand)) {
+        NSDecimalCopy(result, leftOperand);
+        return NSCalculationNoError;
+    }
+
+    NSDecimal left, right;
+
+    // Copy over decimal for structural manipulation.
+    NSDecimalCopy(&left, leftOperand);
+    NSDecimalCopy(&right, rightOperand);
+
+    // normalize both of them
+    NSCalculationError error = NSDecimalNormalize(&left, &right, roundingMode);
+    if (error != NSCalculationNoError) {
+        return error;
+    }
+
+    // If they are both the same sign, then we can directly add the mantissa
+    if ((IS_NEGATIVE(leftOperand) && IS_NEGATIVE(rightOperand)) || (IS_POSITIVE(leftOperand) && IS_POSITIVE(rightOperand))) {
```

Comment:
for some reason I decided to change that from the last CR.

---
In reply to: [80603423](https://github.com/Microsoft/WinObjC/pull/1032#discussion_r80603423) [](ancestors = 80603423)